Issue 13551: loading fixtures in subdirectories

d1rtyf1ng3rs opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | -
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #13551 

An attempt to fix #13551.
Would like hear some comments / critics first, expecially about `getFixturesConfig()` changes.

Commit:
Issue 13551: loading fixtures in subdirectories

samdark commented on framework/console/controllers/FixtureController.php:
```diff
@@ -431,11 +437,18 @@ private function getFixturesConfig($fixtures)
         foreach ($fixtures as $fixture) {
-            $isNamespaced = (strpos($fixture, '\\') !== false);
-            $fullClassName = $isNamespaced ? $fixture . 'Fixture' : $this->namespace . '\\' . $fixture . 'Fixture';
+            // $isNamespaced = (strpos($fixture, '\\') !== false);
```
I think we can check for both `/` and `\` here or normalize everything to linux paths format before the check and the problem will be solved.

d1rtyf1ng3rs replied:
Added one more commit - issue should be solved now.
Added fixture's relative path normalization to linux format in `findFixtures()` method;
reverted changes to `getFixturesConfig()` to its original (so `$isNamespaced` is calculated correctly) and added slash-replace from linux-path-backslashes to namespace slashes.

samdark commented on framework/console/controllers/FixtureController.php:
```diff
@@ -431,11 +437,18 @@ private function getFixturesConfig($fixtures)
-                $config[] = $fullClassName;
+            // istead of guessing if $fixture is namespaced, let's just check it _both_ as non-namespaced and namespaced
+            $className = $fixture . 'Fixture';
+            if (class_exists($className)) {
```
If both namespaced and non-namespaced fixtures with the same base name exist there will be conflict.

Commit:
Issue #13551: fixtures in subdirectories - isNamespaced problem solved

d1rtyf1ng3rs commented:
Added fixture's relative path normalization to linux format in `findFixtures()` method;
reverted changes to `getFixturesConfig()` to its original (so `$isNamespaced` is calculated correctly) and added slash-replace from linux-path-backslashes to namespace slashes.

samdark commented on framework/console/controllers/FixtureController.php:
```diff
@@ -415,7 +415,15 @@ private function findFixtures(array $fixtures = [])
+            // get fixture name relatively to $fixturesPath
+            $relativeName = $relativeDir . basename($fixture, 'Fixture.php');
+            // normalize path to unix format
+            $relativeName = str_replace("\\", "/", $relativeName);
```
Shouldn't it be done to `$fixture` prior to replacements right before `// strip $fixturesPath from $fixture's full path` line?

d1rtyf1ng3rs replied:
Good point. Yes, there might be a problem - `$fixture` is calculated automatically, whereas `$fixturesPath` is provided or may be provided by user. So it's not guaranteed they're in the same format.
I'll provide corresponding changes soon.

samdark commented on framework/console/controllers/FixtureController.php:
```diff
@@ -415,7 +415,15 @@ private function findFixtures(array $fixtures = [])
+            // strip $fixturesPath from $fixture's full path
+            $relativeName = str_replace($fixturesPath . "/", "", $fixture);
+            // get fixtures's directory
+            $relativeDir = dirname($relativeName) == '.' ? '' : dirname($relativeName) . '/';
```
`===` can be used instead of `==`

dynasource commented on framework/console/controllers/FixtureController.php:
```diff
@@ -415,7 +415,15 @@ private function findFixtures(array $fixtures = [])
 
         foreach ($files as $fixture) {
-            $foundFixtures[] = basename($fixture, 'Fixture.php');
+            // strip $fixturesPath from $fixture's full path
```
the added logic should be ensapsulated in a private method with the comments in the description

d1rtyf1ng3rs replied:
> ... unit tests

Roger that; will provide soon.

> ... should be encapsulated in a private method

Why?

dynasource replied:
to see 'what' is does by reading the code instead of reading 'how' it is done

dynasource reviewed (changes requested):
I would like to see your usecase examples covered by unit tests

d1rtyf1ng3rs commented:
> ... unit tests

Roger that; will provide soon.

Commit:
#13551: unit tests added; getRelativeName() private method extracted; path normalization changes

d1rtyf1ng3rs commented:
Welp, it wasn't much of a 'soon', but here it is.
- Unit tests added;
- getRelativeName() method extracted;
- minor path normalization changes.

I have, still, left the inline-comments in getRelativeName() method; feel free to remove them, if those are 'too much'.

dynasource commented on tests/data/console/controllers/fixtures/subdir/SecondFixture.php:
```diff
@@ -0,0 +1,20 @@
+{
+    public function load()
+    {
+        FixtureStorage::$subdirSecondFixtureData[] = 'some data set for subdir/second fixture';
```
I dont see this `string` being used anywhere? Shouldnt it be used in on of the unit tests?

d1rtyf1ng3rs replied:
Well, the string (the value) itself *can*, but *doesn't have to* be used in unit tests.
The property, howerer, [is used in unit-tests](https://github.com/yiisoft/yii2/pull/13553/files/719f6aee4856d2e5bbd9b44100955d9e34f94573#diff-410e9261d60564f099dc41e3dd30a538).
P.s. In fact, all the added fixtures and their usage in tests are *the same* as of existing fixtures.

dynasource replied:
if thats the case, it would be nice to have this string formulated empty or with a more understandable description `'dummy string'`

d1rtyf1ng3rs replied:
> more understandable description 'dummy string'

Agree. But still, I can't see any way this should've been guessed from all the other fixture contents.
I mean, check out any of the `tests/data/console/controllers/fixtures/*Fixture.php` - they all have their data in format of `"some data set for <fixture-name-here> fixture"`...
Hence the question: fixture content should be replaced to `dummy string` only in *added* fixtures, or in *all* of the mentioned fixtures?

dynasource commented on tests/data/console/controllers/fixtures/subdir/FirstFixture.php:
```diff
@@ -0,0 +1,20 @@
+{
+    public function load()
+    {
+        FixtureStorage::$subdirFirstFixtureData[] = 'some data set for first fixture in subdir';
```
I dont see this string being used anywhere? Shouldnt it be used in on of the unit tests?

dynasource commented on framework/console/controllers/FixtureController.php:
```diff
@@ -415,13 +415,36 @@ private function findFixtures(array $fixtures = [])
+        // get fixture name relatively to $fixturesPath
+        $relativeName = $relativeDir . basename($fixture, 'Fixture.php');
+        return $relativeName;
+    }
```
it's hard to determine the quality of the logic without a dedicated unittest. It would be nice to have one for a proper understanding.

d1rtyf1ng3rs replied:
The new function `getFixtureRelativeName()` is triggered on both `load`/`unload` actions, so is involved in all tests that use those actions.
Yet, I agree - its' tests are rather "not obvious" / "non-direct".

If dedicated tests for it are absolutely necessary, I would need some help in form of hints, guidelines or requirements for them. Or any description of what is the "good-enough" form for them.

dynasource replied:
Agree. It's a nice to have as you already covered it.

dynasource commented on framework/console/controllers/FixtureController.php:
```diff
@@ -415,13 +415,36 @@ private function findFixtures(array $fixtures = [])
+     *
+    */
+    private function getFixtureRelativeName($fixture) {
+        // $fixturesPath is normalized to unix format in getFixturesPath()
```
the argument naming does not match the @param doc. Currently I assume an object $fixture being passed.

d1rtyf1ng3rs replied:
Negative.
`$fixture` is a string; one of the strings that `FileHelper::findFiles()` returns [here](https://github.com/yiisoft/yii2/pull/13553/files/719f6aee4856d2e5bbd9b44100955d9e34f94573#diff-96190cbb10ccc75fed657c73f25ae920L414).

dynasource replied:
The function argument should be more clear, in this case $fullFixturePath if the @param comment is correctly formulated.

d1rtyf1ng3rs replied:
Agree. Should be done now.

rob006 commented on framework/console/controllers/FixtureController.php:
```diff
@@ -415,13 +415,36 @@ private function findFixtures(array $fixtures = [])
+     * Calculates $fixture's name relatively to $templatePath.
+     * Basically, strips getFixturePath() and 'Fixture.php' prefix from fixture's full path
+     * @see getFixturePath()
+     * @param $fixture string Full fixture path
```
This phpdoc is invalid - type should go before param name:
```
     * @param string $fixture Full fixture path
```

d1rtyf1ng3rs replied:
Should be fixed now, thanks.

d1rtyf1ng3rs committed:
#13551: minor phpdoc and variable naming changes

cebe committed:
Update FixtureController.php

fixed code style

cebe commented on framework/console/controllers/FixtureController.php:
```diff
@@ -415,13 +415,36 @@ private function findFixtures(array $fixtures = [])
         foreach ($files as $fixture) {
-            $foundFixtures[] = basename($fixture, 'Fixture.php');
+            $relativeName = $this->getFixtureRelativeName($fixture);
+            $foundFixtures[] = $relativeName;
```
no need to have a variable here, just do `$foundFixtures[] = $this->getFixtureRelativeName($fixture);`

cebe commented on framework/console/controllers/FixtureController.php:
```diff
@@ -415,13 +415,36 @@ private function findFixtures(array $fixtures = [])
+        // $fixturesPath is normalized to unix format in getFixturesPath()
+        $fixturesPath = $this->getFixturePath();
+        // normalize $fixture to unix format
+        $fullFixturePath = str_replace("\\", "/", $fullFixturePath);
```
how does this deal with Windows paths like `C:\Somedir`, looks like that does not work to me.

samdark replied:
Windows doesn't really care about which type of slash is used.

cebe replied:
sure, but if the alias resolves to `C:\whatever/fixturepath` it will not match `C:/whatever/fixturepath` as far as I see.

dynasource replied:
sharp. 
We need more unit tests to make ensure this usecase

samdark replied:
@d1rtyf1ng3rs would you please cover these cases?

SilverFire commented:
Merged f111ab441312272250fcd5a4372875715bdd7c66 with minor adjustments in f7c9e6b8528b22d086ea1fd95e0b99f8a3d24e35

Thank you!
